compiler/luci/service/src/CircleShapeInferenceHelper.cpp

```diff
@@ -16,6 +16,104 @@
 
 #include "CircleShapeInferenceHelper.h"
 
+#include <oops/InternalExn.h>
+
+using namespace luci::sinf;
+
+namespace
+{
+
+/**
+ * @brief  Expand shape x and y to same rank by align right and filling with 1
+ */
+void expand_rank(loco::TensorShape &x, loco::TensorShape &y)
+{
+  auto x_rank = x.rank();
+  auto y_rank = y.rank();
+
+  if (x_rank == y_rank)
+    return;
+
+  TensorShapeExpander x_exp(x);
+  TensorShapeExpander y_exp(y);
+
+  auto xy_rank = std::max(x_rank, y_rank);
+
+  x = x_rank > y_rank ? x : x_exp.to(xy_rank);
+  y = y_rank > x_rank ? y : y_exp.to(xy_rank);
+}
+
+/**
+ * @brief  Return shape of expanded dimension of input x and y having same rank
+ */
+loco::TensorShape expand_dimension(const loco::TensorShape &x, const loco::TensorShape &y)
+{
+  assert(x.rank() == y.rank()); // FIX_CALLER_UNLESS
+
+  auto rank = x.rank();
+
+  loco::TensorShape output_shape;
+
+  output_shape.rank(rank);
+
+  // Shape inference rule (commutative)
+  // Values of x, y, and Result are categorized as 1, X, ? where
+  // X is a positive integer greater than 1 and ? is unknown.
+  //     x | y | Result
+  // c1. 1 | X | X
+  // c2. 1 | ? | ?
+  // c3. X | X | X
+  // c4. X | ? | X
+  // c5. ? | ? | ?
+  // Throw exception if none of the below conditions are met
+  // e1. x == y
+  // e2. x != y and x == 1
+  // e3. x != y and y == 1
+  // e4. x is unknown
+  // e5. y is unknown
```

Comment:
This is the main inference logic. I think this kind of comment has to be written for other operators. CC @zetwhite @shs-park

Reply:
Thanks for sharing! This function looks really clear :+1: